### PR TITLE
pythonPackages.tubes: init at 0.2.0

### PR DESCRIPTION
**Pull request by exarkun**
###### Motivation for this change

Tubes is a dependency of Klein 20.6.0.  Klein 20.6.0 has some (but not all) fixes required for compatibility with [Twisted 21.7](https://github.com/NixOS/nixpkgs/pull/132900).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 132944` at 3ac7472e run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>python38Packages.tubes</li>
    <li>python39Packages.tubes</li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/tubes/default.nix:22:3:
```
   |
22 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/duplicate-check-inputs.md">duplicate-check-inputs</a>

Dependencies listed in `propagatedBuildInputs` should not be repeated in `checkInputs`.

Detected duplicates:

- python3.8-Twisted-20.3.0

Near pkgs/development/python-modules/tubes/default.nix:16:3:
```
   |
16 |   checkInputs = [ twisted ];
   |   ^
```

Near pkgs/development/python-modules/tubes/default.nix:14:3:
```
   |
14 |   propagatedBuildInputs = [ characteristic six twisted ];
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/duplicate-check-inputs.md">duplicate-check-inputs</a>

Dependencies listed in `propagatedBuildInputs` should not be repeated in `checkInputs`.

Detected duplicates:

- python3.9-Twisted-20.3.0

Near pkgs/development/python-modules/tubes/default.nix:16:3:
```
   |
16 |   checkInputs = [ twisted ];
   |   ^
```

Near pkgs/development/python-modules/tubes/default.nix:14:3:
```
   |
14 |   propagatedBuildInputs = [ characteristic six twisted ];
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/python-modules/tubes/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  checkInputs = [ twisted ];
+
+  checkPhase = ''
+    trial tubes
+  '';
+
+  meta = with lib; {
+    description = "a data-processing and flow-control engine for event-driven programs";
+    homepage    = "https://github.com/twisted/tubes";
+    license     = licenses.mit;
```
Comment (jonringer) [inline pkgs/development/python-modules/tubes/default.nix:25]: Please add yourself as a maintainer when adding a new package.

https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

File: pkgs/development/python-modules/tubes/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{ lib, buildPythonPackage, fetchPypi
+, characteristic, six, twisted
+}:
+
+buildPythonPackage rec {
+  pname = "Tubes";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tubes/default.nix:6]:
```suggestion
  pname = "tubes";
```

File: pkgs/development/python-modules/tubes/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  version = "0.2.0";
+
+  src = fetchPypi {
+    inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tubes/default.nix:10]:
```suggestion
    pname = "Tubes";
    inherit version;
```

File: pkgs/development/python-modules/tubes/default.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    sha256 = "sha256:0sg1gg2002h1xsgxigznr1zk1skwmhss72dzk6iysb9k9kdgymcd";
+  };
+
+  propagatedBuildInputs = [ characteristic six twisted ];
+
+  checkInputs = [ twisted ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tubes/default.nix:17]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by risicle** [comment]
BTW you don't **have** to do all of these as separate PRs. It's probably ok to just push them all onto the `twisted` PR.

(In some cases it's not _possible_ to do them cleanly as separate PRs because one of the downstream bumps will **only** work with the new version - so don't let this limit you)

File: pkgs/development/python-modules/tubes/default.nix (hunk 4 above)
Comment (exarkun) [reply pkgs/development/python-modules/tubes/default.nix:17]: I thought this should be the case but when I leave twisted out of checkInputs the build fails:

```
running install tests
/nix/store/b39yxip6di5gq3q8513hfvmhfwz4yq77-stdenv-linux/setup: line 1315: trial: command not found
```

Comment (exarkun) [reply pkgs/development/python-modules/tubes/default.nix:18]: I guess it has something to do with python modules vs bin contents.  I can side-step it by using `python -m` though, I guess...

File: pkgs/development/python-modules/tubes/default.nix [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    pname = "Tubes";
+    inherit version;
+    sha256 = "sha256:0sg1gg2002h1xsgxigznr1zk1skwmhss72dzk6iysb9k9kdgymcd";
+  };
+
+  propagatedBuildInputs = [ characteristic six twisted ];
+
+  checkPhase = ''
+    ${python.interpreter} -m twisted.trial -j $NIX_BUILD_CORES tubes
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tubes/default.nix:19]: Please add a pythonImportsCheck and also squash the package related commits.

Comment (exarkun) [reply pkgs/development/python-modules/tubes/default.nix:19]: Thanks.  I did these two things.

**Comment by exarkun** [comment]
I folded this into https://github.com/NixOS/nixpkgs/pull/132900 since that is the motivation, updating multiple packages in one PR is apparently allowed, and it seems like it will be easier to keep all of the pieces in one place.  Thank you for suggesting this @risicle .

**Pull request closed without merging**